### PR TITLE
Enhan jaro cython

### PR DESCRIPTION
**Pull request by SrujithPoondla**
-- Added code for benchmarks in benchmarks directory
-- Added a new directory for benchmark results 
-- Moved cython_levenshtein to new directory py_stringmatching/cython
-- Added cython code for Jaro and Jaro Winkler measures
--Changed unit tests for Generalized Jaccard and Monge Elkan 
--Modified the get_raw_score function in Generalized Jaccard and Monge Elkan to return a floating point number up to 5 decimal places
-- Ran nosetests successfully

### CODE REVIEW

**Comment by paulgc** [comment]
- Remove all .DS_Store files, csv, xlsx files
- Move py_stringmatching/cython to py_stringmatching/similarity_measure/cython

File: py_stringmatching/similarity_measure/jaro.py [hunk 1]
```diff
@@ -50,45 +51,46 @@ def get_raw_score(self, string1, string2):
...
-                        break
-                if string1[i] != string2[j]:
-                    trans_count += 1
-
-        trans_count /= 2
-        common_chars = float(common_chars)
-        weight = ((common_chars / len_s1 + common_chars / len_s2 +
-                   (common_chars - trans_count) / common_chars)) / 3
-        return weight
+        # len_s1 = len(string1)
```
Comment (paulgc) [inline py_stringmatching/similarity_measure/jaro.py]: - remove this commented code

File: py_stringmatching/similarity_measure/jaro_winkler.py [hunk 2]
```diff
@@ -56,19 +57,19 @@ def get_raw_score(self, string1, string2):
...
-        j = min(min_len, 4)
-        i = 0
-        while i < j and string1[i] == string2[i] and string1[i]:
-            i += 1
-
-        if i:
-            jw_score += i * self.prefix_weight * (1 - jw_score)
-
-        return jw_score
+        # jw_score = Jaro().get_raw_score(string1, string2)
```
Comment (paulgc) [inline py_stringmatching/similarity_measure/jaro_winkler.py]: - same applies for this commented code

File: py_stringmatching/tests/test_simfunctions.py [hunk 3]
```diff
@@ -1315,7 +1315,7 @@ def test_set_sim_func(self):
         self.assertAlmostEqual(gj.get_raw_score(['Niall'], ['Neal', 'Njall']), 0.44)
         self.assertEqual(gj.set_sim_func(fn2), True)
         self.assertEqual(gj.get_sim_func(), fn2)
-        self.assertAlmostEqual(gj.get_raw_score(['Niall'], ['Neal', 'Njall']), 0.43333333333333335)
+        self.assertAlmostEqual(gj.get_raw_score(['Niall'], ['Neal', 'Njall']), round(0.43333333333333335,5))
```
Comment (paulgc) [inline py_stringmatching/tests/test_simfunctions.py]: @SrujithPoondla 
- should we just put the rounded similarity score

File: py_stringmatching/utils.py [hunk 4]
```diff
@@ -85,7 +85,7 @@ def convert_bag_to_set(input_list):
 def convert_to_unicode(input_string):
     """Convert input string to unicode."""
     if isinstance(input_string, bytes):
-        return input_string.decode('utf-8')
+        return input_string.decode('utf-8', 'ignore')
```
Comment (paulgc) [inline py_stringmatching/utils.py]: @SrujithPoondla 
- what does the 'ignore' option do? It's better to add comments for future reference.

File: py_stringmatching/cython/cython_jaro_winkler.pyx [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+@cython.wraparound(False)
+
+cdef inline int int_min(int a,int b):
+    if a > b : return b
+    else: return a
+
+def jaro_winkler(unicode string1, unicode string2, float prefix_weight):
+    cdef int i = 0
+
+    cdef float jw_score = Jaro().get_raw_score(string1,string2)
```
Comment (paulgc) [inline py_stringmatching/cython/cython_jaro_winkler.pyx]: @SrujithPoondla 
- should we directly call the jaro cython function ?

File: py_stringmatching/similarity_measure/jaro.py (hunk 1 above)
Comment (SrujithPoondla) [inline py_stringmatching/similarity_measure/jaro.py]: removed the commented code

File: py_stringmatching/similarity_measure/jaro_winkler.py (hunk 2 above)
Comment (SrujithPoondla) [inline py_stringmatching/similarity_measure/jaro_winkler.py]: removed the commented code

File: py_stringmatching/utils.py (hunk 4 above)
Comment (SrujithPoondla) [inline py_stringmatching/utils.py]: Added the comment stating that ignore is added to overcome the UnicodeDecodeError, an issue is also raised on the same. We will look in to this once the cython work is completed.

File: py_stringmatching/cython/cython_jaro_winkler.pyx (hunk 5 above)
Comment (SrujithPoondla) [inline py_stringmatching/cython/cython_jaro_winkler.pyx]: Changed the call to jaro cython function.

**Comment by SrujithPoondla** [comment]
Removed all csv, xlsx and .DS_Store files and also changed the cython directory.

File: py_stringmatching/utils.py [hunk 6]
```diff
@@ -85,8 +86,10 @@ def convert_bag_to_set(input_list):
 def convert_to_unicode(input_string):
     """Convert input string to unicode."""
     if isinstance(input_string, bytes):
-        return input_string.decode('utf-8')
-    return input_string 
+        return input_string.decode('utf-8', 'ignore')  # 'ignore' is added to overcome the UnicodeDecodeError
+        # Issue raised in GitHub to handle or process the data before using
```
Comment (anhaidgroup) [inline py_stringmatching/utils.py]: @SrujithPoondla Mention the issue number in comment

File: py_stringmatching/similarity_measure/cython/cython_jaro.pyx [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+                    k = j + 1
+                    break
+            if string1[i] != string2[j]:
+                trans_count += 1
+    trans_count /= 2
+    cdef float weight = (float(common_chars) / len_str1 + float(common_chars) / len_str2 +
+                         (float(common_chars) - trans_count) / float(common_chars)) / 3
+    return weight
+
+
```
Comment (anhaidgroup) [inline py_stringmatching/similarity_measure/cython/cython_jaro.pyx]: @SrujithPoondla Add comments.

File: py_stringmatching/similarity_measure/cython/cython_jaro_winkler.pyx [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+def jaro_winkler(unicode string1, unicode string2, float prefix_weight):
+    cdef int i = 0
+
+    cdef float jw_score = jaro(string1,string2)
+    cdef int min_len = int_min(len(string1),len(string2))
+
+    cdef int j = int_min(min_len,4)
+
+    while i < j and string1[i] == string2[i] and string1[i]:
```
Comment (anhaidgroup) [inline py_stringmatching/similarity_measure/cython/cython_jaro_winkler.pyx]: @SrujithPoondla add comments

**Comment by anhaidgroup** [comment]
@SrujithPoondla 
- Still a couple of  .DS_Store files are present
- Can you remove the benchmark code from this branch ? You can create a separate branch called 'bench_sim_measures' and add the benchmark code in that branch.

**Pull request closed without merging**